### PR TITLE
create BehaviorTestKit child actors while sync testing

### LINKED ISSUE #24826: Behavior under test cannot stop its child actors
Since the ActorRef returned from the spawn methods is an anonymous TestInbox instance, its path fails the child actor test in StubbedActorContext.scala:122, meaning that `ctx.stop(child)` cannot be contained within Behavior code that is supposed to be tested.

This should be fixed by the first step to #24825.

### PR DESCRIPTION
**Pull request by rkuhn**
This is the first step for #24825 and fixes #24826.

Please disregard the build changes, they are already part of #24827.

### CODE REVIEW

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/EffectfulActorContext.scala [hunk 1]
```diff
@@ -17,30 +17,31 @@ import scala.concurrent.duration.{ Duration, FiniteDuration }
...
     val ref = super.spawnAnonymous(behavior, props)
-    effectQueue.offer(SpawnedAnonymous(behavior, props))
+    effectQueue.offer(SpawnedAnonymous(behavior, props)(ref))
     ref
   }
 
   override def spawnMessageAdapter[U](f: U ⇒ T): ActorRef[U] = {
     val ref = super.spawnMessageAdapter(f)
-    effectQueue.offer(SpawnedAdapter)
+    effectQueue.offer(SpawnedAdapter()(ref))
```
Comment (patriknw) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/EffectfulActorContext.scala]: to make the naming consistent with `SpawnedAnonymous` this should be `SpawnedAnanymousAdapter`

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/EffectfulActorContext.scala [hunk 2]
```diff
@@ -17,30 +17,31 @@ import scala.concurrent.duration.{ Duration, FiniteDuration }
...
     ref
   }
 
   override def spawnMessageAdapter[U](f: U ⇒ T, name: String): ActorRef[U] = {
     val ref = super.spawnMessageAdapter(f, name)
-    effectQueue.offer(SpawnedAdapter)
+    effectQueue.offer(SpawnedNamedAdapter(name)(ref))
```
Comment (patriknw) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/EffectfulActorContext.scala]: to make the naming consistent with `Spawned` this should be `SpawnedAdapter`

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/StubbedActorContext.scala [hunk 3]
```diff
@@ -130,37 +130,41 @@ private[akka] final class StubbedLoggerWithMdc(actual: StubbedLogger) extends Ab
...
-    i.ref
+    val btk = new BehaviorTestKitImpl[U](path / childName.next() withUid rnd().nextInt(), behavior)
+    _children += btk.ctx.self.path.name → btk
+    btk.ctx.self
   }
   override def spawn[U](behavior: Behavior[U], name: String, props: Props = Props.empty): ActorRef[U] =
     _children get name match {
       case Some(_) ⇒ throw untyped.InvalidActorNameException(s"actor name $name is already taken")
       case None ⇒
         // FIXME correct child path for the Inbox ref
```
Comment (patriknw) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/StubbedActorContext.scala]: is this fixed now?

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/javadsl/Effects.scala [hunk 4]
```diff
@@ -20,27 +20,51 @@ object Effects {
...
-  def spawnedAdapter: Effect = SpawnedAdapter
+  def spawnedAdapter(): Effect = SpawnedAdapter[Any]()()
+  /**
+   * The behavior spawned an anonymous adapter, through `ctx.spawnMessageAdapter`
+   */
+  def spawnedAdapter[T](ref: ActorRef[T]): Effect = SpawnedAdapter()(ref)
+  /**
+   * The behavior spawned a named adapter, through `ctx.spawnMessageAdapter`
+   */
+  def spawnedNamedAdapter(name: String): Effect = SpawnedNamedAdapter[Any](name)()
```
Comment (patriknw) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/javadsl/Effects.scala]: spawnedAnonymousAdapter and spawnedAdapter

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/javadsl/TestInbox.scala [hunk 5]
```diff
@@ -4,16 +4,26 @@
...
+
 import scala.collection.JavaConverters._
 import scala.collection.immutable
 
 object TestInbox {
-  def create[T](name: String): TestInbox[T] = new TestInboxImpl(name)
-  def create[T](): TestInbox[T] = new TestInboxImpl[T]("inbox")
+  def create[T](name: String): TestInbox[T] = {
+    val uid = ThreadLocalRandom.current().nextInt()
+    new TestInboxImpl(RootActorPath(Address("akka.actor.typed.inbox", "anonymous")) / name withUid (uid))
```
Comment (patriknw) [inline akka-testkit-typed/src/main/scala/akka/testkit/typed/javadsl/TestInbox.scala]: this address is defined in several places, should we put it in one place

File: akka-testkit-typed/src/test/scala/akka/testkit/typed/scaladsl/BehaviorTestKitSpec.scala [hunk 6]
```diff
@@ -133,14 +149,14 @@ class BehaviorTestKitSpec extends WordSpec with Matchers {
...
-      effects should contain only (Spawned(Child.initial, "child0"), Spawned(Child.initial, "child1", Props.empty))
+      effects should contain only (Spawned(Child.initial, "child0")(null), Spawned(Child.initial, "child1", Props.empty)(null))
     }
 
     "create children when props specified and record effects" in {
       val testkit = BehaviorTestKit[Father.Command](Father.init)
       testkit.run(SpawnChildrenWithProps(2, props))
       val effects = testkit.retrieveAllEffects()
-      effects should contain only (Spawned(Child.initial, "child0", props), Spawned(Child.initial, "child1", props))
+      effects should contain only (Spawned(Child.initial, "child0", props)(null), Spawned(Child.initial, "child1", props)(null))
```
Comment (patriknw) [inline akka-testkit-typed/src/test/scala/akka/testkit/typed/scaladsl/BehaviorTestKitSpec.scala]: why use `(null)` in some places and `()` in others?

File: akka-testkit-typed/src/test/scala/akka/testkit/typed/scaladsl/BehaviorTestKitSpec.scala [hunk 7]
```diff
@@ -203,7 +243,7 @@ class BehaviorTestKitSpec extends WordSpec with Matchers {
       testkit.run(SpawnAndWatchWith("hello"))
       val child = testkit.childInbox("hello").ref
       testkit.retrieveAllEffects() should be(Seq(
-        Spawned(Child.initial, "hello", Props.empty),
+        Spawned(Child.initial, "hello", Props.empty)(),
```
Comment (patriknw) [inline akka-testkit-typed/src/test/scala/akka/testkit/typed/scaladsl/BehaviorTestKitSpec.scala]: I think these `()` looks strange as api. Perhaps we should craft these effect classes without `case class` and have friendly apply and unapply. You don't have to do that in this PR, but let's create a ticket.

**Review by patriknw — COMMENTED** [review]
looking good, some minor things...

File: akka-testkit-typed/src/test/scala/akka/testkit/typed/scaladsl/BehaviorTestKitSpec.scala (hunk 7 above)
Comment (chbatey) [reply akka-testkit-typed/src/test/scala/akka/testkit/typed/scaladsl/BehaviorTestKitSpec.scala]: Or factory methods to match the java dsl

**Review by chbatey — COMMENTED** [review]
Looks like a good approach. I like using the second argument list for capturing the actor ref without breaking equality checks but it does make constructing look a little odd. Maybe add factory methods like the java dsl?

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/EffectfulActorContext.scala (hunk 1 above)
Comment (rkuhn) [reply akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/EffectfulActorContext.scala]: done

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/EffectfulActorContext.scala (hunk 2 above)
Comment (rkuhn) [reply akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/EffectfulActorContext.scala]: done

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/StubbedActorContext.scala (hunk 3 above)
Comment (rkuhn) [reply akka-testkit-typed/src/main/scala/akka/testkit/typed/internal/StubbedActorContext.scala]: yes

File: akka-testkit-typed/src/main/scala/akka/testkit/typed/javadsl/Effects.scala (hunk 4 above)
Comment (rkuhn) [reply akka-testkit-typed/src/main/scala/akka/testkit/typed/javadsl/Effects.scala]: done

File: akka-testkit-typed/src/test/scala/akka/testkit/typed/scaladsl/BehaviorTestKitSpec.scala (hunk 7 above)
Comment (rkuhn) [reply akka-testkit-typed/src/test/scala/akka/testkit/typed/scaladsl/BehaviorTestKitSpec.scala]: done

**Comment by rkuhn** [comment]
I rebased onto current master, looking forward to Jenkins’ green lantern.

**Review by patriknw — APPROVED** [review]
LGTM

**Review by ktoso — APPROVED** [review]
LGTM

**Comment by rkuhn** [comment]
Hooray :-)

**Comment by ktoso** [comment]
:tada:

**Pull request merged**